upgrades/upgrade_testing_framework/steps/step_report_results.py

```diff
@@ -0,0 +1,115 @@
+
+from upgrade_testing_framework.core.framework_step import FrameworkStep
+from upgrade_testing_framework.core.expectation import load_knowledge_base, Expectation
+
+
+class ReportResults(FrameworkStep):
+    """
+    This step reports on the results of each expectation.
+    """
+
+    def _run(self):
+        # Get the state we need
+        eligible_expectation_ids = self.state.eligible_expectations
+        test_output_directory = Path(self.state.get_key('test_results_directory'))
+
+        # Begin the step body
+        path_to_knowledge_base = Path("../knowledge_base")
+        knowledge_base = load_knowledge_base(path_to_knowledge_base)
+        expectations: Dict[str, Expectation] = {e.id: e for e in knowledge_base if e.id in eligible_expectation_ids}
+        test_results: Dict[str, Dict] = parse_robot_results(test_output_directory, eligible_expectation_ids)
+
+        # These three lists will collect tuples of (expectation_id, test_phase). If the expectation was skipped
+        # in all phases, test_phase is be None.
+        passed_expectations: List[tuple] = []
+        skipped_expectations: List[tuple] = []
+        failed_expectations: List[tuple] = []
+
+        for expectation_id, results_by_test_phase in test_results.items():
+            if len(results_by_test_phase) == 0:
+                # No results for any stage were found, this expectation likely doesn't have any tests
+                skipped_expectations.append((expectation_id, None))
+                continue
+            for test_phase, results in results_by_test_phase.items():
+                # The primary assumption being made here is that each expectation should only be in one of
+                # failed, skipped, and passed, and it should greedily be applied to the most "serious" one,
+                # so failed, skipped, passed in that order. That means that if there are multiple assertions
+                # and they have different results (perhaps the first one passes, the second fails, and that
+                # causes the third to be skipped), it will show up as though the entire test failed.
+                if results['fail'] > 0:
+                    failed_expectations.append((expectation_id, test_phase))
+                    continue
+                if results['skip'] > 0:
+                    skipped_expectations.append((expectation_id, test_phase))
+                    continue
+                if results['pass'] > 0:
+                    passed_expectations.append((expectation_id, test_phase))
+                    continue
+        log_results(self.logger, passed_expectations, skipped_expectations, failed_expectations, expectations)
+        self.logger.info("For additional details on test results, see the test "
+                         f"result directory: {test_output_directory.absolute()}")
+
+        # Update our state
+        # N/A
+
+
+def parse_robot_results(test_output_dir: Path, expectation_ids: List[str]) -> Dict[str, Dict]:
```

Comment:
This logic for parsing the xml files doesn't look too intense. Curious though if it makes more sense to use something like the `ExecutionResult` in the robot framework to try and parse.

Reply:
Oh, interesting.  Will take a look at that.